database/ldb/tx.go

```diff
@@ -552,9 +571,65 @@ func (db *LevelDb) DeleteAddrIndex() error {
 	numInBatch := 0
 	for iter.Next() {
 		key := iter.Key()
-		batch.Delete(key)
+		// With a 24-bit index key prefix, 1 in every 2^24 keys is a collision.
+		// We check the length to make sure we only delete address index keys.
+		if len(key) == addrIndexKeyLength {
+			batch.Delete(key)
+			numInBatch++
+		}
+
+		// Delete in chunks to potentially avoid very large batches.
+		if numInBatch >= batchDeleteThreshold {
+			if err := db.lDb.Write(batch, db.wo); err != nil {
+				iter.Release()
+				return err
+			}
+			batch.Reset()
+			numInBatch = 0
+		}
+	}
+	iter.Release()
+	if err := iter.Error(); err != nil {
+		return err
+	}
 
-		numInBatch++
+	batch.Delete(addrIndexMetaDataKey)
+	batch.Delete(addrIndexVersionKey)
+
+	if err := db.lDb.Write(batch, db.wo); err != nil {
+		return err
+	}
+
+	db.lastAddrIndexBlkIdx = -1
+	db.lastAddrIndexBlkSha = wire.ShaHash{}
+
+	return nil
+}
+
+// deleteOldAddrIndex deletes the entire addrindex stored within the DB for a
+// 2-byte addrIndexKeyPrefix. It also resets the cached in-memory metadata about
+// the addr index.
+func (db *LevelDb) deleteOldAddrIndex() error {
+	db.dbLock.Lock()
+	defer db.dbLock.Unlock()
+
+	batch := db.lBatch()
+	defer batch.Reset()
+
+	// Delete the entire index along with any metadata about it.
+	iter := db.lDb.NewIterator(bytesPrefix([]byte("a-")), db.ro)
+	numInBatch := 0
+	for iter.Next() {
+		key := iter.Key()
+		// With a 24-bit index key prefix, 1 in every 2^24 keys is a collision.
+		// We check the length to make sure we only delete address index keys.
+		// We also check the last two bytes to make sure the suffix doesn't
+		// match other types of index that are 34 bytes long.
+		if len(key) == 34 && (key[33] != byte(120) ||
+			(key[32] != byte(116) && key[32] != byte(115))) {
```

Comment:
I agree though I would use a global variable for each suffix to avoid a ton of copies.  Since a string is immutable, go has to make a copy each time you do `[]byte("tx" )` or `[]byte("tx" )` which is hard on the garbage collector.